rdkit

hainm opened the pull request:
not sure how to rebase yet, so just cherry-picked from another branch (rdkit).

hainm commented:
Two examples:

- 
```python
        >>> from rdkit import Chem
        >>> import parmed as pmd
        >>> mol = Chem.MolFromSmiles('Cc1ccccc1')
        >>> struct = pmd.load_rdkit(mol)
```

- 
```python
smile = 'C1=CC=CN=C1'
pmd.rdkit.from_smile(smile)
```

swails commented:
To rebase:

```
git rebase <tree-ish>
```

Where `<tree-ish>` is a branch, commit hash, tag, etc.  What it does is rewind all of the commits that are unique to the current branch you are on until it reaches a common ancestor that is part of the `<tree-ish>` history.  Then it applies all commits from that point to the head of `<tree-ish>`, then it tries to apply the commits made to your branch.  The diagram in `git rebase --help` should help clarify this a bit.  So what I commonly do is this:

```
git rebase origin/master
```

swails commented on parmed/rdkit/__init__.py:
```diff
@@ -0,0 +1,8 @@
+from parmed.rdkit.rdkit import RDKit
+
+load_rdkit = RDKit.load
+from_smile = RDKit.from_smile
```
```from_smiles``` -- it's a SMILES string (the trailing S is part of the acronym).  Does RDKit really use `from_smile`?

hainm replied:
- ah, typos (always think about a 'smile' :D) 

> Does RDKit really use from_smile?

No, It does not have this method. I want to add since I feel frustrated with its (RDKit) syntax (this is understandable RDKit is written in C++ with different interfaces).

swails commented on parmed/rdkit/rdkit.py:
```diff
@@ -0,0 +1,49 @@
+    import StringIO
+except ImportError:
+    import io
+    StringIO = io.StringIO
```
Replace these with the one from `six.moves`:

```
from parmed.utils.six.moves import StringIO
```

swails commented on test/test_parmed_rdkit.py:
```diff
@@ -0,0 +1,29 @@
+    has_rdkit = False
+
+
+@unittest.skipIf(not has_rdkit, "Cannot test load_rdkit module without rdkit.")
```
```
@unittest.skipUnless(has_rdkit, ...
```

swails commented on test/test_parmed_rdkit.py:
```diff
@@ -0,0 +1,29 @@
+        m1 = Chem.MolFromSmiles('C1=CC=CN=C1')
+        parm = pmd.load_rdkit(m1)
+        assert [atom.name for atom in parm.atoms] == ['C1', 'C2', 'C3', 'C4', 'N1', 'C5'] 
+        assert parm.residues[0].name == 'UNL', 'resname must be UNL'
```
Use the `unittest.TestCase` assert methods:

```
self.assertEqual([atom.name for atom in parm.atoms], ...)
self.assertEqual(parm.residues[0].name, 'UNL')
```

swails commented on test/test_parmed_rdkit.py:
```diff
@@ -0,0 +1,29 @@
+        smile = 'C1=CC=CN=C1'
+        parm = pmd.rdkit.from_smile(smile)
+        assert [atom.name for atom in parm.atoms] == ['C1', 'C2', 'C3', 'C4', 'N1', 'C5'] 
+        assert parm.residues[0].name == 'UNL', 'resname must be UNL'
```
Same as above.

swails commented on test/test_parmed_rdkit.py:
```diff
@@ -0,0 +1,29 @@
+        assert parm.residues[0].name == 'UNL', 'resname must be UNL'
+
+if __name__ == '__main__':
+    unittest.main()
```
You can get rid of these lines -- I've been taking them out of the test files.  They were really only useful when I used to support Python 2.4 and I couldn't use nose (I didn't know how to use `unittest` directly at the time).

swails commented:
This looks good, thanks!  I made a bunch of comments.

You'll also have to merge master with this -- it's going to conflict in the listing of the Travis packages.  But it should be a very easy conflict to resolve (just make sure both pysander and rdkit wind up in the installed package list)

hainm commented:
> git rebase origin/master

- got it

> (just make sure both pysander and rdkit wind up in the installed package list)

merged

> from_smiles -- it's a SMILES string

fixed

> from parmed.utils.six.moves import StringIO

fixed

> @unittest.skipUnless(has_rdkit, ...

updated

> Use the unittest.TestCase assert methods:

updated

> Same as above.

updated

> > +if __name__ == '__main__':
> +    unittest.main()
You can get rid of these lines -- I've been taking them out of the test files

removed

Let me know if I missed any of your comments.

swails commented:
pyflakes is failing -- you need to add `rdkit` to `__all__` in `parmed/__init__.py` so pyflakes doesn't think the import is unused.

swails commented on parmed/rdkit/__init__.py:
```diff
@@ -0,0 +1,7 @@
+from parmed.rdkit.rdkit import RDKit
+
+load_rdkit = RDKit.load
+from_smiles = RDKit.from_smiles
```
`__all__ = ['load_rdkit', 'from_smiles']`

swails commented on parmed/rdkit/rdkit.py:
```diff
@@ -0,0 +1,43 @@
+try:
+    from rdkit import Chem
+except ImportError:
+    Chem = None
```
Get rid of these 4 lines and add `from rdkit import Chem` into the function(s) where `Chem` is actually used.  This way you get an `ImportError` with a sensible error message rather than an `AttributeError` complaining about `NoneType` not having a `MolToPDBBlock` or `MolFromSmiles` attribute.  That's what I've started doing.

swails replied:
In addition to giving better tracebacks, it also improves coverage :)

hainm replied:
@swails adding below `class` is ok? 

```python
class RDKit(object):
    from rdkit import Chem

    @staticmethod
    def load(rmol):
```

hainm replied:
nvm, just add to functions.

swails commented:
Looks great, thanks!

hainm commented:
still need to resolve the import error 

> Running test python -m parmed.gromacs._cpp test 2...No module named rdkit

I don't have GROMACS so not sure what's going on here.

swails commented:
> I don't have GROMACS so not sure what's going on here.

This doesn't have anything to do with GROMACS.  `parmed.gromacs._cpp` is a C-preprocessor that has a command-line utility built in.  This is just testing the behavior of the command-line preprocessor (I think it's actually pretty neat :)).

You're going to kick yourself when you see the problem.

`parmed.rdkit` is never getting installed -- you need to add it to the list of installed packages in `setup.py` :smiley:

hainm commented:
> parmed.rdkit is never getting installed

right, parmed.rdkit is a package. I just updated.

(packing in Python is still a myth :)) )

hainm commented on parmed/rdkit/rdkit.py:
```diff
@@ -0,0 +1,41 @@
+        >>> mol = Chem.MolFromSmiles('Cc1ccccc1')
+        >>> struct = pmd.load_rdkit(mol)
+        """
+        from rdkit import Chem
```
this require absolute import since py2 will be confused about 'rdkit' in parmed and rdkit package (from omnia).
(added absolute_import)

hainm commented:
phew, lovely green.

swails commented:
Nice work! Thanks!
